formal-spec/measurements.tex

```diff
@@ -174,6 +177,102 @@ \section{Impact on performance}
 %% the section above.
 %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
+\subsection{Processing Time and Memory Consumption}
+Clearly the most processing intesive task of the update mechanism is the tally phase. It is the phase where all the collected votes are counted in order to reach at a decision for a specific proposal. 
+
+We start with a theoretical time complexity analysis where we assume a worst-case scenario, where we have $n$ participants that all of them vote by submiting a single vote. Also we assume that we have a single proposal, so that within a voting period, the number $n$ of participants coincides to the number of submitted ballots.
+
+In the following we try to break up the operations during the tally phase. In the heart of the tally phase lies the following function call, which is called three times sequencially for each proposal.
+ 
+\begin{lstlisting}[language=Haskell, caption=Tally phase initial function call]
+tallyStake confidence result ballot stakeDistribution adversarialStakeRatio =
+  if stakeThreshold adversarialStakeRatio (totalStake stakeDistribution)
+     <
+     stakeOfKeys votingKeys stakeDistribution
+  then Just result
+  else Nothing
+  where
+    votingKeys = Map.filter (== confidence) ballot
+\end{lstlisting}
+
+\lstinline{Map.filter} is $O(n)$ \footnote{\href{url}{http://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Strict.html\#g:25}}, so \lstinline{votingKeys} is $O(n)$, where $n$ is the number of ballots, which as we have said coincides to the number of participants. At this point, we have a single pass (loop) over $n$ ballots.
```

Comment:
The foot note makes it appear similar to `O(n^2)`. I'd place it after `Map.filter` to avoid confusion.

Reply:
OK